components/feature/push/src/main/java/mozilla/components/feature/push/AutoPushFeature.kt

```diff
@@ -337,6 +335,23 @@ class AutoPushFeature(
     }
 }
 
+internal fun CoroutineScope.launchAndTry(
+    block: suspend CoroutineScope.() -> Unit,
+    errorBlock: (Exception) -> Unit
+): Job {
+    return launch {
+        try {
+            block()
+        } catch (e: RustPushError) {
+            if (e !is CommunicationServerError && e !is CommunicationError) {
```

Comment:
I'd prefer a `when`, but this works. Worst case is that we'll accidentally throw when we shouldn't (e.g. a new "non-terminal" error type is added in the underlying library, and this check degrades), which we should be able to catch in Sentry and resolve.

Perhaps though you can guard against this hypothetical future by forcing an exhaustive `when` treatment? e.g.
```
try {
  block()
} catch (e: RustPushError) {
   if (shouldReThrow(e)) {
     throw e
   }
}

private fun shouldReThrow(e: RustPushError): Boolean {
  return when (e) {
    // Do not re-throw intermittent errors.
    is CommunicationServerError, CommunicationError -> false
    // Re-throw internal errors that are unrecoverable/indicative of bugs. 
    is InternalPanic, CryptoError, AlreadyRegisteredError, StorageError, MissingRegistrationTokenError, StorageSqlError, TranscodingError, UrlParseError -> true
  }
}
```

Reply:
You're right, this is a much safer change. I'd rather fail to compile. I was just being a lazy typist.

Reply:
Well, I did the exact thing as you in my version in FxA... So yeah, we're all lazy :) I'll push a PR to fix up a similar function in FxA.

Reply:
Ah, since the Rust errors are just inherited and not a sealed class, we can never truly make this exhaustive; I have to add _every_ exception to make the compiler happy. 

If we take this direction, we would be swallowing up new error types instead of throwing on them i.e. doing the inverse of the worst case you mentioned.

My preference is to throw on a new error type added. If it was added, that means new functionality was added, and that means the new thing is failing already! 😄 

<img width="272" alt="Kotlin when clause still not happy when we've made every error explicit. We should really ask a-s to change the errors, but that's not really worth their time." src="https://user-images.githubusercontent.com/1370580/68062558-7e668580-fce1-11e9-8c2a-00bfaf07cecd.png">

Reply:
Yeah, I just hit that as well.

> My preference is to throw on a new error type added.

Agreed, let's go with that!